src/wallet/receive.h

```diff
@@ -10,11 +10,55 @@
 #include <wallet/transaction.h>
 #include <wallet/wallet.h>
 
+isminetype InputIsMine(const CWallet& wallet, const CTxIn& txin) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+
+/** Returns whether all of the inputs match the filter */
+bool AllInputsMine(const CWallet& wallet, const CTransaction& tx, const isminefilter& filter);
+
+CAmount OutputGetCredit(const CWallet& wallet, const CTxOut& txout, const isminefilter& filter);
+CAmount TxGetCredit(const CWallet& wallet, const CTransaction& tx, const isminefilter& filter);
+
+bool ScriptIsChange(const CWallet& wallet, const CScript& script) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+bool OutputIsChange(const CWallet& wallet, const CTxOut& txout) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+CAmount OutputGetChange(const CWallet& wallet, const CTxOut& txout) EXCLUSIVE_LOCKS_REQUIRED(wallet.cs_wallet);
+CAmount TxGetChange(const CWallet& wallet, const CTransaction& tx);
+
+CAmount CachedTxGetCredit(const CWallet& wallet, const CWalletTx& wtx, const isminefilter& filter);
```

Comment:
Comparing to 4e11f88320b644b67db55fe737815451ca7d0681 (which I reviewed in #21206) what is the motivation to prefix with `Cached`? Is it relevant for the caller that some caching is involved? Is it a way to remember that concurrent syncs can happen and so the return value can be out of date?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/22100#discussion_r642653123

> Comparing to [4e11f88](https://github.com/bitcoin/bitcoin/commit/4e11f88320b644b67db55fe737815451ca7d0681) (which I reviewed in #21206) what is the motivation to prefix with `Cached`? Is it relevant for the caller that some caching is involved? Is it a way to remember that concurrent syncs can happen and so the return value can be out of date?

There should be no changes since #21206, but `CachedTx` is in debit/credit/change function names when the functions take `CWalletTx` arguments, since the point of `CWalletTx` class is to encapsulate `CTransaction` plus cached balance information. Functions named `Tx` instead of `CachedTx` just take plain `CTransaction` arguments by comparison. And functions named `Input`, `Output`, or `Script` take those things as arguments.

I think it makes calling code more readable if function names explicitly say what they compute values from and whether values may be stale due to caching. But this is just a naming convention I settled on because I didnt like the previous convention where there are so many slightly different functions all having the same name. There could be better approaches I didn't think of though.

Re: concurrency, I think cs_wallet doesn't really allow too much and there shouldn't be caching problems related to that. Caching more affects state transitions when blocks are connected disconnected and transactions are abandoned or replaced or added or removed from the mempool.

Reply:
> Functions named Tx instead of CachedTx just take plain CTransaction arguments by comparison

Thanks for the clarification. An alternative prefix could be `Wallet`, so in this case, could be `WalletTxGetCredit`. But it's fine as is and agree with the convention.

> Re: concurrency, I think cs_wallet doesn't really allow too much and there shouldn't be caching problems related to that. Caching more affects state transitions when blocks are connected disconnected and transactions are abandoned or replaced or added or removed from the mempool.

Right, didn't mean to imply otherwise.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/22100#discussion_r642664678

> > Functions named Tx instead of CachedTx just take plain CTransaction arguments by comparison
> 
> Thanks for the clarification. An alternative prefix could be `Wallet`, so in this case, could be `WalletTxGetCredit`. But it's fine as is and agree with the convention.

That suggestion is pretty good too. Maybe I should `s/CachedTx/WalletTx/g` in this PR. Will do if reviewers want that. I was anticipating renaming `CWalletTx` to `CachedTransaction` later (and namespacing so its full name would be `wallet::CachedTransaction`) but maybe I am jumping the gun a little bit and should stick with `WalletTx` here.
 
> Right, didn't mean to imply otherwise.

Sorry about that, I just misinterpreted "concurrent syncs" it sounds like.

Reply:
Both are just fine to me, as long as the motivation and convention are clear.